node_modules/oae-principals/lib/definitiveDeletion.js

```diff
@@ -0,0 +1,1851 @@
+
+    callback = callback || function(err) {
+        if (err) {
+            log().error({'err': err, 'user': user.id}, 'Unable to send a user a verification email');
+        }
+    };
+
+    if (_.isEmpty(data)) {
+        return callback(null);
+    }
+
+    var userDeleted = {};
+    userDeleted = data[0].userDeleted;
+
+    // Grab the configuration field. This will return the number of months 
+    var month = PrincipalsConfig.getValue(ctx.tenant().alias, 'user', 'delete');
+        
+    // ParseData from variable informations to make an union on user
+    _parseData(data, userDeleted.id, function(err, parseData) {
+
+        if (err) {
+            return callback(err);
+        }
+
+        if (_.isEmpty(parseData)) {
+            return callback(null);
+        }
+
+        var listEmail = [];
+
+        // Send an unique email to each user
+        parseData.users.forEach(function(user, i, arrI) {
+
+            if (!(_.isEmpty(user.userWillBePromoted) && _.isEmpty(user.contentWillBeDeleted) && _.isEmpty(user.userJustLeaving))) {
+
+                // Generate a token if none was specified
+                var token = ShortId.generate();
+
+                // Store the token
+                PrincipalsDAO.storeEmailToken(user.profile.id, user.profile.email, token, function(err) {
+                    if (err) {
+                        return callback(err);
+                    }
+
+                    // The EmailAPI expects a user to have a verified email address. As this is not the case
+                    // when sending an email token, we send in a patched user object
+                    var userToEmail = _.extend({}, user, {'email': user.profile.email});
+
+                    var tenant = TenantsAPI.getTenant(user.profile.tenant.alias);
+
+                    // Send an email to the specified e-mail address
+                    var dataEmail = {
+                        'tenant': ctx.tenant(),
+                        'userDeletedName': userDeleted.displayName,
+                        'resource': user,
+                        'user': userToEmail,
+                        'month': month,
+                        'baseUrl': TenantsUtil.getBaseUrl(ctx.tenant()),
+                        'skinVariables': require('oae-ui').getTenantSkinVariables(ctx.tenant().alias),
+                        'token': token
+                    };
+
+                    // We pass the current date in as the 'hashCode' for this email. We need to be able to send
+                    // a copy of the same email for the 'resend email token' functionality. As we don't expect
+                    // that this logic will get stuck in a loop this is probably OK
+                    EmailAPI.sendEmail('oae-principals', 'notify', userToEmail.profile, dataEmail, {'hash': Date.now()});
+                    listEmail.push(user.profile.email);
+
+                    // Callback usefull for tests
+                    if (i === arrI.length-1) { 
```

Comment:
Because it's a foreach loop I have to keep these two returns ... Or I could make a for loop instead of a foreach loop : 
```
for (var i = 0; i<=user.lenght; i++) {
    if (i === user.lenght) {
        return callback(null, listEmail);
    }
    ...
}
```